File: src/main/java/com/github/svegon/capi/event/render/SameFluidCheckCallback.java
package com.github.svegon.capi.event.render;

import net.fabricmc.fabric.api.event.Event;
import net.fabricmc.fabric.api.event.EventFactory;
import net.minecraft.fluid.FluidState;
import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable;

public interface SameFluidCheckCallback {
    Event<SameFluidCheckCallback> EVENT = EventFactory.createArrayBacked(SameFluidCheckCallback.class,
            (rendered, neighbor, callback) -> {}, listeners -> (rendered, neighbor, callback) -> {
        for (SameFluidCheckCallback listener : listeners) {
            listener.isSameFluid(rendered, neighbor, callback);

            if (callback.isCancelled()) {
                return;
            }
        }
    });

    void isSameFluid(FluidState rendered, FluidState neighbor, CallbackInfoReturnable<Boolean> callback);
}

File: src/main/kotlin/io/github/svegon/mclientapi/event/render/SameFluidCheckCallback.kt
package io.github.svegon.capi.event.render

import io.github.svegon.capi.event.render.SameFluidCheckCallback
import net.fabricmc.fabric.api.event.Event
import net.fabricmc.fabric.api.event.EventFactory
import net.minecraft.fluid.FluidState
import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable
import java.util.function.Function
import kotlin.Array
import kotlin.Boolean
import kotlin.invoke

interface SameFluidCheckCallback {
    fun isSameFluid(rendered: FluidState?, neighbor: FluidState?, callback: CallbackInfoReturnable<Boolean?>?)

    companion object {
        @JvmField
        val EVENT: Event<SameFluidCheckCallback?> = EventFactory.createArrayBacked(
            SameFluidCheckCallback::class.java,
            SameFluidCheckCallback { rendered: FluidState?, neighbor: FluidState?, callback: CallbackInfoReturnable<Boolean?>? -> },
            Function { listeners: Array<SameFluidCheckCallback?>? ->
                SameFluidCheckCallback { rendered: FluidState?, neighbor: FluidState?, callback: CallbackInfoReturnable<Boolean?>? ->
                    for (listener in listeners) {
                        listener.isSameFluid(rendered, neighbor, callback)

                        if (callback.isCancelled()) {
                            return@SameFluidCheckCallback
                        }
                    }
                }
            })
    }
}
